arkhn/fhir-river
# Hapi loader

## Description
This PR aims to provide a new "load" implementation for fhir-river ETL while targeting the HAPI-FHIR-JPA persistence layer.
It uses hapi-fhir backend abstractions (which use the hibernate ORM under the hood) to interact with FHIR resources and the database (and ES).

Almost all the code comes from [hapi-fhir-jpaserver-starter](https://github.com/hapifhir/hapi-fhir-jpaserver-starter) except `KafkaTopicConfig` `KafkaConsumerConfig` and `ResourceConsumer` (main class).

## Technical details
Right now, the implementation is a bit different than the current one. The kafka consumer(s) listen for events and buffer them until a given buffer size is reached (I hardcoded 100 but i'll make this configurable): when this happens, a transaction is sent to the db (in a separate thread in order not to block the consumer(s)).
Note that kafka consumers can be scaled (using `concurrency = "8"` in the `@KafkaListener` annotation) and seems to work well as long as the topic is correctly partitioned (1 partition per listener).

## Tests
Integration tests:
- spawn postgres, ES and kafka instances
- run an instance of the hapi-loader kafka consumer
- send an event on the listened topic containing a serialized resource
- make sure it has been updated or created in the db
